fossas/fossa-cli
# Ane 891 exclude test deps

# Overview

This PR adds support for excluding test dependencies in the new go tactic and outputting pseudo versions in a format the fossa core can understand.

This works by altering the algorithm used from one that linearly goes through the output of `go list` to one that walks the dependency tree from each main package and skips over test deps. Technically this code works by skipping over test dependencies rather than graphing and removing them later. I found that when I graphed the test deps I would occasionally find that a package had test dependencies that didn't have an entry in `go list` that corresponded to it. That is a fatal error that would cause the Go strategy to fall back to the old dynamic or static tactic. I only encountered this with test dependencies, so for the time being I've opted to not graph test deps at all.

## Acceptance criteria

* Test dependencies and their transitive deps should be excluded.
* Pseudo versions should be changed into the appropriate hash and work when uploaded to fossa core.

## Testing plan

1. Build fossa from this branch.
2. Run using `fossa analyze --experimental-v3-go-resolution --revision test-exclusions`
3. Run again using a version of fossa from master: `fossa analyze --experimental-v3-go-resolution --revision no-test-exclusions`
4. Notice if there are differences in deps.

As one of my test repos, I used https://github.com/ipfs/kubo. The version of fossa on the master branch contains the module `github.com/davecgh/go-spew`, however the version on this branch does not.

Using this command in the `kubo` directory we can see that there is just one package in that module used in the project:
```
> go list -json -deps -e all | jq 'select(."Module"."Path" == "github.com/davecgh/go-spew")'
{
  "Dir": "/home/go/pkg/mod/github.com/davecgh/go-spew@v1.1.1/spew",
  "ImportPath": "github.com/davecgh/go-spew/spew",
  "Name": "spew",
  "Doc": "Package spew implements a deep pretty printer for Go data structures to aid in debugging.",
  "Root": "/home/go/pkg/mod/github.com/davecgh/go-spew@v1.1.1",
  "Module": {
    "Path": "github.com/davecgh/go-spew",
    "Version": "v1.1.1",
...
```
If we search for packages that import that one, there is just one:
```
>  go list -json -e -deps all | jq 'select(."Imports"[] | test("github.com/davecgh/go-spew/spew"))' 
{
  "Dir": "/home/go/pkg/mod/github.com/stretchr/testify@v1.8.1/assert",
  "ImportPath": "github.com/stretchr/testify/assert",
...
```
Which is in turn imported by `github.com/stretchr/testify/require` which is imported by `go-libp2p/p2p/net/swarm/testing`.
This last dep doesn't appear in any other package's import list:
```
> go list -json -e -deps all | jq 'select(."Imports"[] | test("github.com/libp2p/go-libp2p/p2p/net/swarm/testing"))'
<nothing here>
```
But does appear in `TestImports`:
```
>  go list -json -e -deps all | jq 'select(."TestImports"[] | test("github.com/libp2p/go-libp2p/p2p/net/swarm/testing"))' 
{
  "Dir": "/home/go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.21.1/internal/net",
  "ImportPath": "github.com/libp2p/go-libp2p-kad-dht/internal/net",
  "Name": "net",
...
```

## Risks

No risks specific to this PR, other than the fact that it's trying to support Go which has been difficult for us in the past.

## References

[ANE-891](https://fossa.atlassian.net/browse/ANE-891)

## Checklist



[ANE-891]: https://fossa.atlassian.net/browse/ANE-891?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
